clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java

```diff
@@ -2421,6 +2413,34 @@ public void testGetOffsetsFencedLeaderEpoch() {
         assertEquals(0L, metadata.timeToNextUpdate(time.milliseconds()));
     }
 
+    @Test
+    public void testGetOffsetByTimeWithPartitionsRetryCouldTriggerMetadataUpdate() {
+        buildFetcher();
+        List<Errors> retriableErrors = Arrays.asList(Errors.NOT_LEADER_FOR_PARTITION,
+            Errors.REPLICA_NOT_AVAILABLE, Errors.KAFKA_STORAGE_ERROR, Errors.OFFSET_NOT_AVAILABLE,
+            Errors.LEADER_NOT_AVAILABLE, Errors.FENCED_LEADER_EPOCH, Errors.UNKNOWN_LEADER_EPOCH);
+
+        for (Errors retriableError : retriableErrors) {
+            subscriptions.assignFromUser(singleton(tp0));
+            client.updateMetadata(initialUpdateResponse);
+            assertEquals(1, metadata.fetch().nodes().size());
+
+            Map<String, Integer> partitionNumByTopic = new HashMap<>();
+            partitionNumByTopic.put(topicName, 1);
+            final int updatedNodeSize = 3;
```

Comment:
This an awkward way to verify that the metadata gets updated. Nothing in the test actually involves the use of the new metadata. A more realistic scenario would be a leader change. Could we let the metadata update indicate a leader change and then use `prepareResponseFrom` to ensure that the retry goes to a particular node?